### DISCUSSION TITLE
How does Nginx handle load balancing when proxy_pass is a domain?

### DISCUSSION THREAD

**Discussion by manedurphy** [nginx/nginx#391] [General]
I have included a simplified version of the configuration I am working with. You can see that incoming requests to `my-service.domain.com` are forwarded to `http://my-service:8081`. In this context, `my-service` is a container running in a Docker Swarm along with the Nginx server. Nginx is just a proxy. The issue I am facing is that when the service for `my-service` is updated, the old containers are shutdown and new ones are started. When new containers for `my-service` are started, they are assigned different IP addresses. When an request comes in for `my-service.domain.com` after the `my-service` containers have been updated and assigned new IPs, Nginx will still include the old IPs in the load balancing rotation, so sometimes the request a will timeout when an old IP is selected in the rotation which of course adds latency to the request. I've implemented a workaround where a script will automatically restart the Nginx containers when the `my-service` containers are assigned new IPs, but ideally Nginx would be able to figure out that `my-service` no longer resolves to the old IPs.

```nginx
# HTTPS server configuration for my-service.domain.com
server {
    listen 443 ssl;
    http2 on;
    server_name my-service.domain.com;

    ssl_certificate /etc/ssl/certs/my-service.domain.com.crt;
    ssl_certificate_key /etc/ssl/private/my-service.domain.com.key;

    location / {
        proxy_pass http://my-service:8081;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }
}
```

**Comment by HanadaLee** [comment]
you can try one of these solutions.

tips: you can set valid=xxx for resolver to ensure that dns records can be updated in a timely manner.

solutions 1
```
# HTTPS server configuration for my-service.domain.com
server {
    listen 443 ssl;
    http2 on;
    server_name my-service.domain.com;

    ssl_certificate /etc/ssl/certs/my-service.domain.com.crt;
    ssl_certificate_key /etc/ssl/private/my-service.domain.com.key;

    resolver 127.0.0.11;
    location / {
        set $upstream my-service;
        proxy_pass http://$upstream:8081;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }
}
```

solutions 2 (only available after nginx version 1.27.3)
```
# HTTPS server configuration for my-service.domain.com
upstream upstream-service {
  resolver 127.0.0.11;
  server my-service resolve;
}

server {
    listen 443 ssl;
    http2 on;
    server_name my-service.domain.com;

    ssl_certificate /etc/ssl/certs/my-service.domain.com.crt;
    ssl_certificate_key /etc/ssl/private/my-service.domain.com.key;

    location / {
        proxy_pass http://upstream-service:8081;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }
}
```

**Reply by manedurphy** [reply]
I have sandboxed this issue outside of the Docker Swarm and have verified the functionality of both presented solutions. The documentation makes is clear that the [resolve](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#server:~:text=as%20permanently%20unavailable.-,resolve,-monitors%20changes%20of) keyword is designed to handle dynamic IPs for server domain names. Although I confirmed the functionality of solution 1, I don't quite understand why this configures Nginx to dynamically update the IPs for a service's domain. Is there something special about setting the name of the service in a variable and passing that variable to `proxy_pass`? @HanadaLee can you provide some clarify on this? And if this stated somewhere in the documentation, I'd like to see that as well. Appreciated 🤝

**Comment by HanadaLee** [comment]
This is related to nginx parsing the upstream name in proxy_pass. If the upstream name is a string, it will be parsed at startup and a shadow upstream block will be created (an automatically created upstream block).

The feature of the upstream block is that domain name resolution is performed only at startup (until version 1.27.3 supports the use of the resolve parameter).

However, if the upstream name in the proxy_pass directive is a variable, since nginx cannot obtain the specific upstream during the configuration parsing phase, it will not create a new shadow upstream block, but will use the resolve directive to obtain the upstream IP and initiate a proxy request during operation (at the cost of being unable to maintain a long connection).

The documentation of the proxy_pass directive introduces this feature, but does not explain the principle in detail.
https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_pass
```
Parameter value can contain variables. In this case, if an address is specified as a domain name, the name is searched among the described server groups, and, if not found, is determined using a [resolver](https://nginx.org/en/docs/http/ngx_http_core_module.html#resolver).
```

**Comment by manedurphy** [comment]
@HanadaLee what do you mean by `at the cost of being unable to maintain a long connection`? My interpretation of this statement is that when a variable is passed into `proxy_pass`, the DNS resolution happens just before the request is passed on to the upstream server and will close the TCP connection as soon as HTTP finishes. Does this mean that `keep-alive` is irrelevant in that case, and that every new HTTP request will open an new TCP connection when the value passed to `proxy_pass` is a variable?

**Reply by HanadaLee** [reply]
Yes, when you use a variable as the upstream name to proxy to the upstream, nginx will close the upstream connection after each request.  This is a huge burden for a high-load proxy server.

Only when you use the upstream block or a clear domain name and configure the keepalive related instructions (keepalive in upstream block and proxy_set_header Connection "";) can you ensure that the upstream connection is not closed.

Before nginx 1.27.3, in order to solve this problem, I used openresty's balancer_by_lua_block to achieve dynamic upstream domain name resolution and maintain upstream long connections at the same time.

Fortunately, with the release of nginx 1.27.3, nginx officially ported the upstream dynamic resolution function of nginx plus from nginx plus to nginx oss. Now you can implement dynamic upstream domain name resolution while nginx is running in this version, and you can also configure the keepalive and load balancing algorithms.
https://github.com/nginx/nginx/commit/db6870e06dde7ab249e9a41a0e0a76219f82dd8c

**Reply by manedurphy** [reply]
Excellent, I appreciate the additional information. Thanks again for your assistance.

**Reply by Maryna-f5** [reply]
Thanks a lot, @HanadaLee, for your help!
